### PR TITLE
[Fix] Generalize spec-V1 hidden_states slicing to all V1 algorithms (stacked on #26150)

### PR DESCRIPTION
**Pull request by javierlimt6**
## Motivation

`return_hidden_states=True` under any speculative-decoding V1 algorithm silently returns fewer per-token hidden states than `completion_tokens` (issue #26163). The non-spec path is correct; the spec-V1 path was not. PR #26150 fixed the NGRAM-only case. This PR generalizes that fix to the remaining V1 algorithms (EAGLE, EAGLE3, multi-layer EAGLE, FROZEN_KV_MTP).

**Root cause:** `process_batch_result_decode` did `req.hidden_states.append(logits_output.hidden_states[i]...)`, which assumes `hidden_states` is shape `[batch_size, hidden_dim]`. After spec verify, the worker gathers `hidden_states` down to accepted positions via `accept_indices`, producing a flat `[sum(num_accept_per_req), hidden_dim]` tensor in batch-row order. `hidden_states[i]` then picks the i-th flat row across all requests instead of request i's slice — so the returned hidden states are silently mis-attributed and undercount completion_tokens.

**Reproduction** (from #26163, on `main` pre-fix, commit `885274196`):

```python
import sglang as sgl
llm = sgl.Engine(
    model_path="meta-llama/Llama-2-7b-chat-hf",
    speculative_algorithm="EAGLE",
    speculative_draft_model_path="lmsys/sglang-EAGLE-llama2-chat-7B",
    speculative_num_steps=5, speculative_eagle_topk=8, speculative_num_draft_tokens=64,
    enable_return_hidden_states=True, mem_fraction_static=0.7,
)
outputs = llm.generate(
    ["Repeat: the quick brown fox the quick brown fox the quick brown fox"],
    sampling_params={"temperature": 0, "max_new_tokens": 32},
    return_hidden_states=True,
)
hs = outputs[0]["meta_info"]["hidden_states"]
ct = outputs[0]["meta_info"]["completion_tokens"]
print(f"len(hs)={len(hs)}  completion_tokens={ct}  (expect equal)")
# Observed: len(hs)=17  completion_tokens=32
```

### Stacked on #26150

This PR is **stacked on top of #26150** — its 4 commits (truthiness fix by @Nas01010101, offset-builder + NGRAM regression test by @kpham-sgl) sit at the bottom of this branch. My 5 commits sit on top. The current diff intentionally includes #26150's hunks so reviewers can run CI against a consistent state.

**Once #26150 merges into `main`, I will rebase to drop those commits** and the diff will collapse to ~88 lines (18 production + 70 CI test). Keeping the branch in draft until that rebase is done.

### Production (`python/sglang/srt/managers/scheduler_components/batch_result_processor.py`)

1. **Drop the `is_ngram()` gate** in `process_batch_result_decode` so the per-request offset path runs for any V1 algorithm that populates both `logits_output.hidden_states` and `result.num_correct_drafts_per_req_cpu`. Rename `ngram_hs_offsets` → `spec_v1_hs_offsets` for honesty.
2. **Length-invariant runtime check** so any future drift between `hidden_states.shape[0]` and `sum(num_correct_drafts + 1)` raises immediately instead of silently mis-attributing rows. Uses `if ... raise RuntimeError(...)` (not `assert` — survives `python -O` / `PYTHONOPTIMIZE=1`).
3. Updated the comment to reflect that this branch now handles all V1 spec algorithms, not just NGRAM.

### Tests

- New CI regression test at `test/registered/spec/eagle/test_eagle_hidden_states.py` for EAGLE V1, mirroring the NGRAM test pattern from #26150. Stage `extra-a` + runner `1-gpu-large` matches existing EAGLE test convention.

### Algorithm coverage table

| Algorithm | Fixed | Notes |
|---|---|---|
| NGRAM | ✅ (via #26150) | |
| EAGLE / EAGLE3 | ✅ this PR | Same `EagleVerifyOutput` layout |
| multi-layer EAGLE | ✅ this PR | Same flat layout (`multi_layer_eagle_worker.py:582`) |
| FROZEN_KV_MTP | ✅ this PR | `FrozenKVMTPVerifyOutput = EagleVerifyOutput` literally |
| STANDALONE | ➖ pre-existing limitation | Uses `CaptureHiddenMode.NULL` on target verify (`eagle_worker.py:812-815`), so `hidden_states` is `None` and never enters this branch. STANDALONE + `return_hidden_states` was already unsupported. |
| DFLASH | ➖ pre-existing limitation | Deliberately nulls `hidden_states` in `dflash_info.py:495`. |
| All V2 algorithms | ❌ separate follow-up | Different strided layout `[bs * num_draft_tokens, H]`, different code branch in `process_batch_result_decode`. |

## Accuracy Tests

**Not applicable.** This PR does not modify model forward, attention, kernels, sampling, or the accept/reject decisions in spec verify. It only fixes the bookkeeping of per-token `hidden_states` returned to the user via `meta_info["hidden_states"]`. Generated tokens (`output_ids`), logprobs, accept counts, and `completion_tokens` are bit-identical to pre-fix behavior.

The regression test asserts the invariant directly:
```python
self.assertEqual(len(hs), out["meta_info"]["completion_tokens"])
```

## Speed Tests and Profiling

**Not applicable.** The added work is:
- One O(batch_size) Python cumsum loop per decode step (a list-append loop over `num_correct_drafts_per_req_cpu`)
- One scalar comparison for the runtime invariant check

Both are negligible relative to the existing per-decode-step costs (target forward, draft forward, verify kernel, KV cache management). The hot-path slice itself (`hidden_states[start:end].cpu().clone().tolist()`) was already present in #26150's NGRAM-only path — this PR only changes which algorithms enter it.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests) — `test/registered/spec/eagle/test_eagle_hidden_states.py`.
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations) — no user-facing documentation changes needed; the fix is internal.
- [x] Provide accuracy and speed benchmark results — N/A (no model output / no perf impact, see sections above).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance) — naming follows `.claude/rules/speculative-naming.md` (`num_correct_drafts` excludes bonus, `+1` adds bonus per Rule 3; verb-form `accept` not `accepted` per Rule 1).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process). **Please wait until #26150 merges first**, then I will rebase and flip out of draft.
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) — the new EAGLE test lives in stage `extra-a`, so `/tag-run-ci-label extra` is needed to exercise it on H100 hardware.
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

## Out of scope (worth follow-up)

- **`_maybe_collect_customized_info`** (`batch_result_processor.py:171`) has the same `v[i]` bug pattern this PR fixes for `hidden_states`. Latent for any plugin/feature using per-token-flat `customized_info` fields. Separate PR.
- **V2 algorithms** need an analogous fix on the V2 branch (different layout).

---
### CI States

Latest PR Test (Base): :x: [Run #26364272506](https://github.com/sgl-project/sglang/actions/runs/26364272506)
Latest PR Test (Extra): :x: [Run #26364272477](https://github.com/sgl-project/sglang/actions/runs/26364272477)

### CODE REVIEW

**Comment by javierlimt6** [comment]
Thanks for the review @gemini-code-assist — both suggestions applied in [b284c0c3d](https://github.com/sgl-project/sglang/pull/26217/commits/b284c0c3d):

- Hoisted `logits_output.hidden_states.tolist()` out of the per-request loop into the spec-V1 block. For a batch with N requests with `return_hidden_states=True`, this replaces N per-row `.cpu().tolist()` calls with one full-tensor `.tolist()` — eliminates N-1 GPU→CPU syncs.
- Dropped `.clone()` from both spec-V1 slice sites.

Scoped the hoist to the spec-V1 branch only (kept conditional on `is_spec_v1 and num_correct_drafts_per_req_cpu is not None`) so non-spec / V2 paths are untouched in this PR — the same optimization would apply to the non-spec path at L644 but that's out of scope here.
